bench/src/simulated_network.rs

```diff
@@ -0,0 +1,250 @@
+use std::collections::VecDeque;
+use std::io::{Error, IoSliceMut};
+use std::net::{IpAddr, Ipv6Addr, SocketAddr};
+use std::sync::{Arc, Mutex};
+use std::task::{Poll, Waker};
+use std::time::{Duration, Instant};
+
+use quinn::udp::{RecvMeta, Transmit, UdpState};
+use quinn::AsyncUdpSocket;
+
+use queue::InboundQueue;
+
+#[derive(Debug)]
+pub struct InMemorySocketHandle {
+    pub network: Arc<InMemoryNetwork>,
+    pub addr: SocketAddr,
+}
+
+impl AsyncUdpSocket for InMemorySocketHandle {
+    fn poll_send(
+        &self,
+        _state: &UdpState,
+        _cx: &mut std::task::Context,
+        transmits: &[Transmit],
+    ) -> Poll<Result<usize, Error>> {
+        let now = Instant::now();
+        for transmit in transmits {
+            let transmit = Transmit {
+                destination: transmit.destination,
+                ecn: transmit.ecn,
+                contents: transmit.contents.clone(),
+                src_ip: Some(self.addr.ip()),
+                segment_size: transmit.segment_size,
+            };
+
+            self.network.send(now, self.addr, transmit);
+        }
+
+        if transmits.is_empty() {
+            Poll::Pending
+        } else {
+            Poll::Ready(Ok(transmits.len()))
+        }
+    }
+
+    fn poll_recv(
+        &self,
+        cx: &mut std::task::Context,
+        bufs: &mut [IoSliceMut<'_>],
+        meta: &mut [RecvMeta],
+    ) -> Poll<std::io::Result<usize>> {
+        let socket = self.network.socket(self.addr);
+        let mut inbound = socket.inbound.lock().unwrap();
+
+        let max_transmits = meta.len();
+        let mut received = 0;
+
+        let out = meta.iter_mut().zip(bufs);
+        for (in_transit, (meta, buf)) in inbound.receive(max_transmits).zip(out) {
+            received += 1;
+            let transmit = in_transit.transmit;
+
+            // Meta
+            meta.addr = in_transit.source_addr;
+            meta.ecn = transmit.ecn;
+            meta.dst_ip = Some(transmit.destination.ip());
+            meta.len = transmit.contents.len();
+            meta.stride = transmit.segment_size.unwrap_or(meta.len);
+
+            // Buffer
+            buf[..transmit.contents.len()].copy_from_slice(&transmit.contents);
+        }
+
+        if received == 0 {
+            if inbound.is_empty() {
+                // Store the waker so we can be notified of new transmits
+                let mut waker = socket.waker.lock().unwrap();
+                if waker.is_none() {
+                    *waker = Some(cx.waker().clone())
+                }
+            } else {
+                // Wake up next time we can read
+                let next_read = inbound.time_of_next_receive();
+                let waker = cx.waker().clone();
+                tokio::task::spawn(async move {
+                    tokio::time::sleep_until(next_read.into()).await;
+                    waker.wake();
+                });
```

Comment:
If we assume at most one reader, which I think is reasonable, we could store the `Sleep` future in `self` and poll it directly.